org.eclipse.gef/src/org/eclipse/gef/commands/CompoundCommand.java

```diff
@@ -142,21 +150,23 @@ public Object[] getChildren() {
 	/**
 	 * @return the List of contained Commands
 	 */
-	public List getCommands() {
+	public List<? extends Command> getCommands() {
```

Comment:
This now forces users to store the return value in a `List<? extends Command>`. If it just returns a `List<Command>`, then the user can store it as either one of those.

Reply:
The reason I'm typically using this pattern is that I've seen users doing in similar situations the following:

`List<MyCommand> lst = (List<MyCommand>) compound.getCommnds()`

Which is not possible if we return a `List<Command>`. But I'm not sure which of the both is more common and has more impact.

Reply:
Hm, fair point. But that brings up another possibility: Users already can do something like:
`List<MyCommand> lst = compound.getCommands()`
Which now causes compile errors. With the <? extends Command>, the user has at least the ability to fix this compile error, so I would then say that's the preferable option.

Reply:
> But I'm not sure which of the both is more common and has more impact.

As an FYI: I just had a look at how we use GEF in our other product and we always use one of the following patterns:
```
for (Object o :  compound.getCommands()) {
    Command command = (Command)o;
    ...
}
```
```
List<Command> commands = compound.getCommands();
```
At least on our side, the second one will require code adaptations once we update, but I'm not sure how representative a sample size of one is going to be 😅.
But I think your point still stands. With the <? extends Command>, it will always be possible to resolve the compile errors in user code with minimal effort.

Reply:
This nicely shows why it is so hard to clean the code. 

But not cleaning is IMHO also not an option. I already found one place where the documentation was not matching the implementation. With Java generics this is now clearer. I hope our users will see it the same way.